pkg/dockerregistry/server/pullthroughmanifestservice.go

```diff
@@ -51,27 +51,39 @@ func (m *pullthroughManifestService) Get(ctx context.Context, dgst digest.Digest
 func (m *pullthroughManifestService) remoteGet(ctx context.Context, dgst digest.Digest, options ...distribution.ManifestServiceOption) (distribution.Manifest, error) {
 	dcontext.GetLogger(ctx).Debugf("(*pullthroughManifestService).remoteGet: starting with dgst=%s", dgst.String())
 	image, rErr := m.imageStream.GetImageOfImageStream(ctx, dgst)
+	var ref reference.DockerImageReference
 	if rErr != nil {
 		switch rErr.Code() {
-		case imagestream.ErrImageStreamNotFoundCode, imagestream.ErrImageStreamImageNotFoundCode:
+		case imagestream.ErrImageStreamNotFoundCode:
 			dcontext.GetLogger(ctx).Errorf("remoteGet: unable to get image %s in imagestream %s: %v", dgst.String(), m.imageStream.Reference(), rErr)
 			return nil, distribution.ErrManifestUnknownRevision{
 				Name:     m.imageStream.Reference(),
 				Revision: dgst,
 			}
+		case imagestream.ErrImageStreamImageNotFoundCode:
+			ref, rErr = m.imageStream.ResolveUpstreamRef(ctx, dgst)
+			if rErr != nil {
+				dcontext.GetLogger(ctx).Errorf("remoteGet: unable to get image %s in imagestream %s: %v", dgst.String(), m.imageStream.Reference(), rErr)
+				return nil, distribution.ErrManifestUnknownRevision{
+					Name:     m.imageStream.Reference(),
+					Revision: dgst,
+				}
+			}
 		case imagestream.ErrImageStreamForbiddenCode:
 			dcontext.GetLogger(ctx).Errorf("remoteGet: unable to get access to imagestream %s to find image %s: %v", m.imageStream.Reference(), dgst.String(), rErr)
 			return nil, distribution.ErrAccessDenied
+		default:
+			return nil, rErr
 		}
-		return nil, rErr
-	}
-
-	ref, err := reference.Parse(image.DockerImageReference)
-	if err != nil {
-		dcontext.GetLogger(ctx).Errorf("bad DockerImageReference (%q) in Image %s@%s: %v", image.DockerImageReference, m.imageStream.Reference(), dgst.String(), err)
-		return nil, err
+	} else {
```

Comment:
why the `else`? it's already confusing to do a `if a != b {} else {}` (instead of `if a == b {} else {}`).
and in this case the `swtich` inside the `if` block has a default which returns, so when `if rErr != nil` is true won't the function always return, thus making the `else` unnecessary?

Reply:
nevermind, I realise that not every `case` inside the previous switch returns so we do need the `else` here